### ISSUE TITLE
Connected but not connected

### ISSUE THREAD

**Issue by RossBoylan** [HenrikBengtsson/ucsf-vpn#40]
I got the series of inconsistent indications about my connection status, as shown in the bottom.  I was unable to connect for awhile, but the situation somehow resolved and I am connected now.

1. connected via vpn (first command below) but then did not use the connection for about an hour.  
2. Then tried to RDP and got a message it couldn't reach the remote machine, usually an indication the connection is down.  
3. `ucsf-vpn status` said I was not connected (though it did report an openconnect process running).
4. Yet a few seconds later when I tried to `start` it said I was already connected.  
5. Ran `stop`, seemingly without event.  
6. `start` then said "ERROR: Internet connection is not working".
7. `log` reported no errors.
8.  `start` 9 minutes later succeeded.  My RDP connection succeeded.

Speculation: if the connection is inactive it gets torn down, either on my side or the server.  But this isn't a failure mode ucsf-vpn is familiar with.

Additionally, my connection has been unstable, periodically reporting it is disconnected and then automatically reconnecting, usually succeeding before timeout.  The problem there may be on my end.

```{bash}
root@barley:~# date; ~ross/src/ucsf-vpn/bin/ucsf-vpn start  # 1 in the list above
Wed 28 Oct 2020 01:11:32 PM PDT
NOTE: Open the Duo Mobile app on your smartphone or tablet to confirm ...
OK: OpenConnect status: 'openconnect' process running (PID=8066)
DEBUG: running list-records in /etc/resolvconf/run/interface: enp5s0.inet
lo.named
tun0
OK: Public IP information: ip=128.218.42.217, hostname=, org=AS5653 University of California San Francisco
OK: Connected to the VPN

root@barley:~# date; ~ross/src/ucsf-vpn/bin/ucsf-vpn status # 3
Wed 28 Oct 2020 02:13:44 PM PDT
OpenConnect status: 'openconnect' process running (PID=8066)
Public IP information: ip=, hostname=, org=
Not connected to the VPN

root@barley:~# date; ~ross/src/ucsf-vpn/bin/ucsf-vpn start # 4
Wed 28 Oct 2020 02:13:56 PM PDT
WARNING: Skipping - already connected to the VPN
OK: OpenConnect status: 'openconnect' process running (PID=8066)
ERROR: Public IP information: ip=, hostname=, org= [ucsf-vpn 5.2.0-9000, OpenConnect 8.02-1+deb10u1]

root@barley:~# date; ~ross/src/ucsf-vpn/bin/ucsf-vpn stop # 5
Wed 28 Oct 2020 02:17:10 PM PDT
DEBUG: running list-records in /etc/resolvconf/run/interface: enp5s0.inet
lo.named
OK: OpenConnect status: No 'openconnect' process running
OK: Public IP information: ip=66.181.128.6, hostname=m128-6.dsl.rawbw.com, org=AS7961 Raw Bandwidth Communications, Inc.
OK: Not connected to the VPN

root@barley:~# date; ~ross/src/ucsf-vpn/bin/ucsf-vpn start # 6
Wed 28 Oct 2020 02:17:37 PM PDT
ERROR: Internet connection is not working [ucsf-vpn 5.2.0-9000, OpenConnect 8.02-1+deb10u1]

root@barley:~# date; ~ross/src/ucsf-vpn/bin/ucsf-vpn log # 7
Wed 28 Oct 2020 02:26:03 PM PDT
root@barley:~# date; ~ross/src/ucsf-vpn/bin/ucsf-vpn --verbose start # 8
Wed 28 Oct 2020 02:26:19 PM PDT
INFO: Pinging '9.9.9.9' once
INFO: Verified that internet connection works
INFO: Getting public IP (from https://ipinfo.io/ip)
INFO: Pinging '9.9.9.9' once
INFO: Preparing to connect to VPN server 'remote.ucsf.edu'
INFO: Administrative ("sudo") rights already establish
NOTE: Open the Duo Mobile app on your smartphone or tablet to confirm ...
INFO: Connecting to VPN server 'remote.ucsf.edu'
INFO: Connected to VPN server
INFO: validate='pid,ipinfo'
OK: OpenConnect status: 'openconnect' process running (PID=11699)
INFO: Pinging '9.9.9.9' once
DEBUG: running list-records in /etc/resolvconf/run/interface: enp5s0.inet
lo.named
tun0
INFO: Verified that internet connection works
INFO: Getting public IP (from https://ipinfo.io/ip)
OK: Public IP information: ip=128.218.43.45, hostname=, org=AS5653 University of California San Francisco
OK: Connected to the VPN
```

**Comment by HenrikBengtsson** [comment]
# Reproducible example

I can reproduce the problem off reconnecting to the VPN after temporarily losing my internet connection. It toggle the airplane mode of my notebook by using `nmcli radio all off` and `nmcli radio all on`.

## Working VPN connection

I start out with a working VPN connection:

```sh
$ ucsf-vpn status
OpenConnect status: 'openconnect' process running (PID=1545143)
Connected to the VPN

$ ucsf-vpn details
{
  "ip": "128.218.*.*",
  "city": "San Francisco",
  "region": "California",
  "country": "US",
  "loc": "37.7749,-122.4194",
  "org": "AS5653 University of California San Francisco",
  "postal": "94102",
  "timezone": "America/Los_Angeles",
  "readme": "https://ipinfo.io/missingauth"
}

$ ping -W 3 -c 1 9.9.9.9
PING 9.9.9.9 (9.9.9.9) 56(84) bytes of data.
64 bytes from 9.9.9.9: icmp_seq=1 ttl=54 time=15.9 ms

--- 9.9.9.9 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 15.879/15.879/15.879/0.000 ms

$ ucsf-vpn routing | head -n 6
Default non-VPN network interface: wlp0s20f3
IP routing table (126 entries):
default dev tun0 scope link 
default via 192.168.1.1 dev wlp0s20f3 proto dhcp metric 600 
3.7.35.0/25 via 192.168.1.1 dev wlp0s20f3 src 192.168.1.19 
3.21.137.128/25 via 192.168.1.1 dev wlp0s20f3 src 192.168.1.19
```

## Toggling internet connection breaks VPN connection

Then I quickly turn off and on my wireless internet connection (I don't have a wired one):

```sh
$ nmcli radio all off && nmcli radio all on
```

After this, I see:

```sh
$ ucsf-vpn status
OpenConnect status: 'openconnect' process running (PID=1545143)
Connected to the VPN

$ ucsf-vpn details
ERROR: Internet connection is not working [ucsf-vpn 5.7.0-9012, OpenConnect 8.20-1]

$ ping -W 3 -c 1 9.9.9.9
PING 9.9.9.9 (9.9.9.9) 56(84) bytes of data.

--- 9.9.9.9 ping statistics ---
1 packets transmitted, 0 received, 100% packet loss, time 0ms
```

Here `ucsf-vpn status` is clearly wrong, because it only checks if the OpenConnect process is still running.

Interestingly, if we look at `ucsf-vpn routing` (develop branch), which basically report on `ip route show`, we see that it lost almost all of it's IP routing rules, except for `default dev tun0 scope link` and `10.48.93.92 dev tun0 scope link`.

```diff
--- stopped.txt	2024-05-06 13:26:59.416933196 -0700
+++ broken.txt	2024-05-06 13:26:30.152776276 -0700
@@ -1,7 +1,9 @@
 Default non-VPN network interface: wlp0s20f3
-IP routing table (5 entries):
+IP routing table (7 entries):
+default dev tun0 scope link 
 default via 192.168.1.1 dev wlp0s20f3 proto dhcp metric 600 
 10.0.3.0/24 dev lxcbr0 proto kernel scope link src 10.0.3.1 linkdown 
+10.48.*.* dev tun0 scope link 
 169.254.0.0/16 dev wlp0s20f3 scope link metric 1000 
 x.x.0.0/16 dev docker0 proto kernel scope link src x.x.0.1 linkdown 
 192.168.1.0/24 dev wlp0s20f3 proto kernel scope link src 192.168.1.19 metric 600 
```


## Stopping the VPN connection fixes the internet connection

If I do:

```sh
$ ucsf-vpn stop
OK: OpenConnect status: No 'openconnect' process running
OK: Not connected to the VPN
```

it will take a few seconds before it completes, but immediately after than, I have working internet connection again;

```sh
$ ping -W 3 -c 1 9.9.9.9
PING 9.9.9.9 (9.9.9.9) 56(84) bytes of data.
64 bytes from 9.9.9.9: icmp_seq=1 ttl=56 time=4.78 ms

--- 9.9.9.9 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 4.775/4.775/4.775/0.000 ms

$ ucsf-vpn status
OpenConnect status: No 'openconnect' process running
Not connected to the VPN
```

```sh
$ bin/ucsf-vpn routing
Default non-VPN network interface: wlp0s20f3
IP routing table (5 entries):
default via 192.168.1.1 dev wlp0s20f3 proto dhcp metric 600 
10.0.3.0/24 dev lxcbr0 proto kernel scope link src 10.0.3.1 linkdown 
169.254.0.0/16 dev wlp0s20f3 scope link metric 1000 
x.x.0.0/16 dev docker0 proto kernel scope link src x.x.0.1 linkdown 
192.168.1.0/24 dev wlp0s20f3 proto kernel scope link src 192.168.1.19 metric 600
```


## Restarting the VPN connection is the quick fix

After having a broken internet connection, the VPN connection can be fixed by `ucsf vpn restart`;

```sh
$ nmcli radio all off && nmcli radio all on

$ bin/ucsf-vpn restart
Enter 'push' (default), 'phone', 'sms', a 6 or 7 digit token, or press your YubiKey: <push>
NOTE: Open the Duo Mobile app on your smartphone or tablet to confirm ...
OK: OpenConnect status: 'openconnect' process running (PID=1564979)
OK: Connected to the VPN

$ ping 9.9.9.9
PING 9.9.9.9 (9.9.9.9) 56(84) bytes of data.
64 bytes from 9.9.9.9: icmp_seq=1 ttl=54 time=9.03 ms
^C
--- 9.9.9.9 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 9.030/9.030/9.030/0.000 ms
```

**Comment by HenrikBengtsson** [comment]
I've updated `ucsf vpn status` to run all validation methods, so that we now get:

```sh
$ nmcli radio all off && nmcli radio all on
$ bin/ucsf vpn status
ERROR: Internet connection is not working [ucsf-vpn 5.7.0-9014, OpenConnect 8.20-1]
ERROR: Internet connection is not working [ucsf-vpn 5.7.0-9014, OpenConnect 8.20-1]
ERROR: Internet connection is not working [ucsf-vpn 5.7.0-9014, OpenConnect 8.20-1]
ERROR: Internet connection is not working [ucsf-vpn 5.7.0-9014, OpenConnect 8.20-1]
OpenConnect status: 'openconnect' process running (PID=251248)
Public IP information: ip=, hostname=, org=
ERROR: Conflicting results whether connected to the VPN. This can happen if the network dropped temporarily
while on the VPN. You might be able to fix it with 'ucsf vpn restart' [ucsf-vpn 5.7.0-9014, OpenConnect 8.20-1]
```

whenever we break our VPN connection, e.g. due to a network/internet glitch.

**Comment by HenrikBengtsson** [comment]
I'll consider this one resolved. Please open a new issue, if not working as expected.
